### PR TITLE
Allow int varref with other numeric operands in a multiplicative expression

### LINKED ISSUE #32940: `additive-expr` with operands of different basic types does not give the correct compilation error when the contextually expected type of the `additive-expr` is `decimal`
**Description:**
$title

**Steps to reproduce:**
```ballerina
public function main(decimal a, int b) {
    decimal a1 = a + b; // gives `incompatible types: expected 'decimal', found 'int'`,
                        // but should be, `operator '+' not defined for 'decimal' and 'int'`
    decimal? a2 = a + b; // gives `operator '+' not defined for 'decimal' and 'int'`, as expected
}
```
**Affected Versions:**
slbeta3-rc3

### LINKED ISSUE #35179: Allow int variable reference with other numeric operands in a multiplicative expression
Please refer to : https://github.com/ballerina-platform/ballerina-spec/issues/261

- [x] Open github discussion
- [ ] Implement SymbolTable changes and add tests

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$title

Fixes #35179
Fixes #32940

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by KavinduZoysa** [comment]
@rdulmina, shall we add test cases for int max min and decima max min.

**Review by KavinduZoysa — APPROVED** [review]

**Comment by SandaruJayawardana** [comment]
I think we have to preserve commutative law
```ballerina
int x1 = 9223372036854775807;
int x2 = 2;
float y1 = 0.1 * x1 * x2; // 1.84467440737095526E18

float y2 = x1 * x2 * 0.1; // error: {ballerina}NumberOverflow {"message":"int range overflow"} at test2:main(test2.bal:8)
```

**Comment by rdulmina** [comment]
> I think we have to preserve commutative law
> 
> ```ballerina
> int x1 = 9223372036854775807;
> int x2 = 2;
> float y1 = 0.1 * x1 * x2; // 1.84467440737095526E18
> 
> float y2 = x1 * x2 * 0.1; // error: {ballerina}NumberOverflow {"message":"int range overflow"} at test2:main(test2.bal:8)
> ```

This is a general issue eg below a sample of float multiplication also gives different outputs
```bal
float x1 = 92233720e300f;
float x2 = 10;
float y1 = 0.1 * x1 * x2; // 9.223372E307

float y2 = x1 * x2 * 0.1; // Infinity
```

We need to think about all the possible cases and handle this. Created a separate issue for this #35706.

**Review by gimantha — APPROVED** [review]

**Review by SandaruJayawardana — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/division-operation.bal [hunk 1]
```diff
@@ -294,6 +294,457 @@ function testDivisionNullable() {
...
+
+    var g = constDecimal / b;
+    decimal? var5 = g;
+    assertEqual(var5, 4.1d);
+
+    var i = constDecimal / constInt;
+    decimal? var7 = i;
+    assertEqual(var7, 4.1d);
+}
+
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/division-operation.bal:747]: Tests when the type of the lhs operand is inferred as `float` or `decimal`
```ballerina
    var a = 12d;
    int b = 10;
    var x = a / b;

    var c = 12f;
    var y = c / b;
```

**Pull request merged**